File: app/src/main/java/com/tryden/leetcode/java/easy/ValidPalindrome.java
package com.tryden.leetcode.java.easy;

/*
 *
 * [125] Valid Palindrome
 *
 * https://leetcode.com/problems/valid-palindrome/description/
 *
 */

public class ValidPalindrome {

    public static void main(String[] args) {
        String s = "A man, a plan, a canal: Panama";
        int x = 128;
        isPalindrome(x);
        isPalindrome(s);
    }
    public static Boolean isPalindrome(int x) {
        if (x<0 || (x!=0 && x%10==0)) return false;
        int rev = 0;
        while (x>rev){
            rev = rev*10 + x%10;
            x = x/10;
        }
        return (x==rev || x==rev/10);

    }

    public static Boolean isPalindrome(String s) {
        if (s.isEmpty()) return true;

        int head = 0, tail = s.length()-1;
        char cHead, cTail;
        while(head <= tail) {
            cHead = s.charAt(head);
            cTail = s.charAt(tail);
            if (!Character.isLetterOrDigit(cHead)) {
                head++;
            } else if (Character.isLetterOrDigit(cTail)){
                tail--;
            } else {
                if (Character.toLowerCase(cHead) != Character.toLowerCase(cTail)) {
                    return false;
                }
                head++;
                tail--;
            }
        }
        return true;
    }
}

File: app/src/main/java/com/tryden/leetcode/kotlin/easy/ValidPalindrome.kt
package com.tryden12.leetcode_kotlin

/*
 *
 * [125] Valid Palindrome
 *
 * https://leetcode.com/problems/valid-palindrome/description/
 *
 */

    fun main() {

    }

    // Elegant
    fun isPalindrome(s: String) = s.filter {
        it.isLetterOrDigit() }
        .run { equals(reversed(), true) }


    // Map
    fun isPalindrome2(s: String) : Boolean {
        val result = s.filter { it.isLetterOrDigit() }
            .map { it.lowercaseChar() }
        return result == result.reversed()
    }
